### ISSUE TITLE
"validate_in_current_graph" breaks Dim tag in gRPC worker thread

### ISSUE THREAD

**Issue by patrick-wilken** [rwth-i6/returnn#1040]
I'm using a [gRPC server](https://grpc.github.io/grpc/python/grpc.html#create-server), where I setup a RETURNN Engine on startup (creating Tensorflow session + graph) and run inference on incoming requests (which are run as new threads, to my understanding). This used to work without problems.

With recent RETURNN versions I get this error: `dynamic time axis dim (None) (axis 1) but size_placeholder {} misses information`.

I tracked the issue down to `Dim._validate_in_current_graph()` which contains:
```
g = tf_util.get_root_graph()
if tf_util.get_root_graph(tensor.graph) is not g:  # maybe from an earlier run which reuses the dim tag
  # Reset and cleanup.
  self.dyn_size_ext = None
  same_base = self.get_same_base()
  same_base._same_for_batch_ctx.pop((self.batch, self.control_flow_ctx), None)
  self.batch = None  # it is invalid in the new graph
  self.control_flow_ctx = None  # also invalid
```

In my case the two Graph objects are different and `self.dyn_size_ext` gets reset, leading to the error. It seems that `tf_compat.v1.get_default_graph()` when called from the worker tread is different to `tensor.graph`, which probably points to the graph of the main thread where the tensor was created. Not sure whether this is expected and whether this can be fixed.

In any case, resetting the Dim is dangerous, isn't it? I only get to that part of the code in the worker thread because I call `layer.output.get_placeholder_as_batch_major()` to create the fetches dict. That this has side effects on the Dim tags is certainly unexpected. Especially given that Dim tags are used globally (see `def __copy__(self): retunn self`) and should therefore be immutable.

For what do we need `validate_in_current_graph` actually? "maybe from an earlier run which reuses the dim tag", so to make it possible for the user to define dim tags and use them for multiple model runs in the same python session?

**Comment by albertz** [comment]
This is a very old change already. You have not updated in the last 1-2 years or so? You should also better watch the changes being made.

Anyway:

Why don't you just set the default graph correctly? In all threads (I think it is thread local). It's `Graph.as_default()`.

RETURNN (just as TF itself) assumes in multiple places that the default graph is the correct and relevant graph. This is just one example now where it breaks if this is not correct. But I think it's a valid assumption and also should not really be any problem to just set the default graph to whatever graph you want.

The code is there because the session/graph might have been reset (e.g. pretraining, new epoch, or whatever) and in `Dim` objects we keep tensor references and it breaks when we use the old ones from the old graph.

**Comment by patrick-wilken** [comment]
It's from September. I follow changes on the master, but I don't update code in production often where this was relevant.

Alright, `with engine.tf_session.graph.as_default():` around feed dict creation code seems reasonable and solves it.

Just for reference: `tf.compat.v1.get_default_graph()` returns different graphs in different threads, but so far I did not have to care because in the worker I do not alter the graph, and the session (referenced from the main thread) still has the right graph.

**Comment by albertz** [comment]
> Just for reference: `tf.compat.v1.get_default_graph()` returns different graphs in different threads

Yes this is what I said above: I think it is thread local. But this should not really be a problem. Just define it as the default graph in all threads where you use it.
